TELCODOCS-756-AWS: First draft (#53022)
openshift/openshift-docs: modules/nvidia-gpu-aws-adding-a-gpu-node.adoc

could be scaled to add a node in a particular region and zone.
+
[source,terminal]
----
$ oc get machines -n openshift-machine-api | grep worker
----
+
.Example output
+
[source,terminal]
----
preserve-dsoc12r4-ktjfc-worker-us-east-2a-dts8r      Running   m5.xlarge   us-east-2   us-east-2a   3d11h
preserve-dsoc12r4-ktjfc-worker-us-east-2b-dkv7w      Running   m5.xlarge   us-east-2   us-east-2b   3d11h
preserve-dsoc12r4-ktjfc-worker-us-east-2b-k58cw      Running   m5.xlarge   us-east-2   us-east-2b   3d11h
----

. Make a copy of one of the existing compute `MachineSet` definitions and output the result to a JSON file by running the following command.
This will be the basis for the GPU-enabled worker machine set definition.
+
[source,terminal]
----
$ oc get machineset preserve-dsoc12r4-ktjfc-worker-us-east-2a -n openshift-machine-api -o json > <output_file.json>
----

. Edit the JSON file and make the following changes to the new `MachineSet` definition:
+
* Replace `worker` with `gpu`. This will be the name of the new machine set.
* Change the instance type of the new `MachineSet` definition to `g4dn`, which includes an NVIDIA Tesla T4 GPU.

> To learn more about AWS `g4dn` instance types, see https://aws.amazon.com/ec2/instance-types/#Accelerated_Computing[Accelerated Computing].

Reviewer: the docs.openshift.com transform is happy with this, but I don't trust it not to trip the portal transform :sweat_smile: 

```suggestion
To learn more about AWS `g4dn` instance types, see link:https://aws.amazon.com/ec2/instance-types/#Accelerated_Computing[Accelerated Computing].
```

----
+
.Example output
+
[source,terminal]
----
10c10

< "name": "preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a",
---
> "name": "preserve-dsoc12r4-ktjfc-worker-us-east-2a",

21c21

< "machine.openshift.io/cluster-api-machineset": "preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a"
---
> "machine.openshift.io/cluster-api-machineset": "preserve-dsoc12r4-ktjfc-worker-us-east-2a"

31c31

< "machine.openshift.io/cluster-api-machineset": "preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a"
---
> "machine.openshift.io/cluster-api-machineset": "preserve-dsoc12r4-ktjfc-worker-us-east-2a"

60c60

< "instanceType": "g4dn.xlarge",
---
> "instanceType": "m5.xlarge",
----

. Create the GPU-enabled compute machine set from the definition by running the following command:
+
[source,terminal]
----
$ oc create -f preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a.json
----
+
.Example output
+
[source,terminal]
----
machineset.machine.openshift.io/preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a created
----

.Verification

. View the machine set you created by running the following command:
+
[source,terminal]
----
$ oc -n openshift-machine-api get machinesets | grep gpu
----
+

> The MachineSet replica count is set to `1` so a new Machine object is created automatically.

Reviewer: ```suggestion
The `MachineSet` replica count is set to `1` so a new `Machine` object is created automatically.
```

// Module included in the following assemblies:
//

> //  * installing/installing-nvidia-gpu-aws.adoc

Reviewer: Update assembly reference

----
$ oc get machinesets -n openshift-machine-api
----
+
.Example output
+
[source,terminal]
----
NAME                                        DESIRED   CURRENT   READY   AVAILABLE   AGE
preserve-dsoc12r4-ktjfc-worker-us-east-2a   1         1         1       1           3d11h
preserve-dsoc12r4-ktjfc-worker-us-east-2b   2         2         2       2           3d11h
----

. At this time, there is only one worker machine per machine set, though a machine set
could be scaled to add a node in a particular region and zone.
+
[source,terminal]
----
$ oc get machines -n openshift-machine-api | grep worker
----
+
.Example output
+
[source,terminal]
----
preserve-dsoc12r4-ktjfc-worker-us-east-2a-dts8r      Running   m5.xlarge   us-east-2   us-east-2a   3d11h
preserve-dsoc12r4-ktjfc-worker-us-east-2b-dkv7w      Running   m5.xlarge   us-east-2   us-east-2b   3d11h
preserve-dsoc12r4-ktjfc-worker-us-east-2b-k58cw      Running   m5.xlarge   us-east-2   us-east-2b   3d11h
----

. Make a copy of one of the existing worker MachineSet definitions and output the result to a ``JSON`` file.
This will be the basis for the GPU-enabled worker machine set definition.
+
[source,terminal]
----

> oc get machineset preserve-dsoc12r4-ktjfc-worker-us-east-2a -n openshift-machine-api -o json > <output_file.json>

Reviewer: ```suggestion
$ oc get machineset preserve-dsoc12r4-ktjfc-worker-us-east-2a -n openshift-machine-api -o json > <output_file.json>
```

----

> . Edit the JSON file and make the following changes to the new MachineSet definition:

Reviewer: ```suggestion
. Edit the JSON file and make the following changes to the new `MachineSet` definition:
```

+
* Replace ``worker`` with ``gpu``. This will be the name of the new machine set.

> * Change the instance type of the new MachineSet definition to ``g4dn``,

Reviewer: ```suggestion
* Change the instance type of the new `MachineSet` definition to ``g4dn``,
```

which includes an NVIDIA Tesla T4 GPU.

> To learn more about AWS g4dn instance types, see

Reviewer: ```suggestion
To learn more about AWS `g4dn` instance types, see
```

https://aws.amazon.com/ec2/instance-types/#Accelerated_Computing[Accelerated Computing].
+
[source,terminal]
----
$ jq .spec.template.spec.providerSpec.value.instanceType preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a.json

"g4dn.xlarge"
----
+
The ``<output_file.json>`` file is saved as ``preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a.json``.

 . Update the following fields in ``preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a.json``:
+

> * ``.metadata.name`` to a name containing ``gpu``.

Reviewer: Just needs the string `gpu` in there somewhere, not more specific?

* ``.spec.selector.matchLabels["machine.openshift.io/cluster-api-machineset"]`` to
match the new ``.metadata.name``.

* ``.spec.template.metadata.labels["machine.openshift.io/cluster-api-machineset"]``
to match the new ``.metadata.name``.

* ``.spec.template.spec.providerSpec.value.instanceType`` to ``g4dn.xlarge``.

> . Perform a ``diff`` of the original worker definition and the new GPU-enabled node definition.

Reviewer: This is a verification step, right? Might be good to point that out, e.g.:
```suggestion
. To verify your changes, perform a ``diff`` of the original compute definition and the new GPU-enabled node definition by running the following command:
```

+
[source,terminal]
----
oc -n openshift-machine-api get preserve-dsoc12r4-ktjfc-worker-us-east-2a -o json | diff preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a.json -
----
+
.Example output
+
[source,terminal]
----
10c10

< "name": "preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a",
---
> "name": "preserve-dsoc12r4-ktjfc-worker-us-east-2a",

21c21

< "machine.openshift.io/cluster-api-machineset": "preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a"
---
> "machine.openshift.io/cluster-api-machineset": "preserve-dsoc12r4-ktjfc-worker-us-east-2a"

31c31

< "machine.openshift.io/cluster-api-machineset": "preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a"
---
> "machine.openshift.io/cluster-api-machineset": "preserve-dsoc12r4-ktjfc-worker-us-east-2a"

60c60

< "instanceType": "g4dn.xlarge",
---
> "instanceType": "m5.xlarge",
----

> . Create the new MachineSet from the definition:

Reviewer: ```suggestion
. Create the GPU-enabled compute machine set from the definition by running the following command:
```

+
[source,terminal]
----
$ oc create -f preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a.json
----
+
.Example output
+
[source,terminal]
----
machineset.machine.openshift.io/preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a created
----

> . View the new MachineSet:

Reviewer: ```suggestion
. View the machine set you created by running the following command:
```

+
[source,terminal]
----
$ oc -n openshift-machine-api get machinesets | grep gpu
----
+
The MachineSet replica count is set to ``1`` so a new Machine object is created automatically.

+
.Example output
+
[source,terminal]
----
preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a   1         1         1       1           4m21s
----

> . View the new Machine object:

Reviewer: ```suggestion
. View the `Machine` object that the machine set created by running the following command:
```

+
[source,terminal]
----
$ oc -n openshift-machine-api get machines | grep gpu
----
+
.Example output
+
[source,terminal]
----
preserve-dsoc12r4-ktjfc-worker-gpu-us-east-2a    running    g4dn.xlarge   us-east-2   us-east-2a  4m36s

> ----

Reviewer: Could move these two steps under a `.Verification` section

> Note that there is no need to specify a namespace for the node. The node definition is cluster scoped.

Reviewer: This note seems like it might belong with a specific step above?

> // Module included in the following assemblies:

Reviewer 2: This module is good but it's not for GPU-enablement when you install you cluster.  It's more for GPU-enabled node after you've already installed your cluster. 

What you need to do is add it to the end of the assembly on creating Machines/MachineSets for AWS: https://docs.openshift.com/container-platform/4.11/machine_management/creating_machinesets/creating-machineset-aws.html

As an example, Google Cloud Platform already has its post-install Machine/MachineSet config in the docs: https://docs.openshift.com/container-platform/4.11/machine_management/creating_machinesets/creating-machineset-gcp.html#machineset-gcp-enabling-gpu-support_creating-machineset-gcp

We just need to do the same thing for the AWS instructions.

Author: Moved **nvidia-gpu-aws-adding-a-gpu-node.adoc** to the **creating-machineset-aws.adoc** assembly.

+
[source,terminal]
----
$ oc get machinesets -n openshift-machine-api
----
+
.Example output
+
[source,terminal]
----
AME DESIRED CURRENT READY AVAILABLE AGE

openshift-blog-txxtf-worker-us-east-2a 1 1 1 1 22m

openshift-blog-txxtf-worker-us-east-2b 1 1 1 1 22m

openshift-blog-txxtf-worker-us-east-2c 1 1 1 1 22m
----

. At this time, there is only one worker machine per MachineSet, though a MachineSet
could be scaled to add a node in a particular region and zone.
+
[source,terminal]
----
$ oc get machines -n openshift-machine-api | grep worker
----
+
.Example output
+
[source,terminal]
----
openshift-blog-txxtf-worker-us-east-2a-8grnb running m4.large us-east-2 us-east-2a 51m

openshift-blog-txxtf-worker-us-east-2b-vw4ph running m4.large us-east-2 us-east-2b 51m

openshift-blog-txxtf-worker-us-east-2c-qpd4q running m4.large us-east-2 us-east-2c 51m
----

. Make a copy of one of the existing worker
MachineSet definitions and output the result to a JSON file.
This will be the basis for the GPU-enabled worker machine set definition.
+
[source,terminal]
----
$ oc get machineset openshift-blog-txxtf-worker-us-east-2a -n openshift-machine-api -o json >

> ----

Reviewer 3: Need to specify a file name for redirected machineset json output file after the ">".

Author: Added <output_file.json>

. Edit the JSON file and make the following changes to the new MachineSet definition:
+
* Replace ``worker`` with ``gpu``. his will be the name of the new machine set.
* Change the instance type of the new MachineSet definition to ``p3``,

> which includes an NVIDIA Tesla V100 GPU.

Reviewer 3: Replace "p3" with "g4dn".  change "v100" to "T4".

> To learn more about AWS P3 instance types, see

Reviewer 3: change "P3" to "g4dn"

https://aws.amazon.com/ec2/instance-types/#Accelerated_Computing[Accelerated Computing].
+
[source,terminal]
----
$ jq .spec.template.spec.providerSpec.value.instanceType openshift-blog-txxtf-gpu-us-east-2a.json

"p3.2xlarge"
----

. Change the name and self link to a unique name that identifies the new
MachineSet. Delete the status section from the definition.
Perform a diff of the original worker definition and the new GPU-enabled node definition.
+
[source,terminal]
----
$ oc -n openshift-machine-api get machineset openshift-blog-txxtf-worker-us-east-2a -o json | diff openshift-blog-txxtf-gpu-us-east-2a.json -
----
+
.Example output
+
[source,terminal]
----
10c10

< "name": "openshift-blog-txxtf-gpu-us-east-2a", --- > "name": "openshift-blog-txxtf-worker-us-east-2a",

13c13

< "selfLink": "/apis/machine.openshift.io/v1beta1/namespaces/openshift-machine-api/machinesets/openshift-blog-txxtf-gpu-us-east-2a", --- > "selfLink": "/apis/machine.openshift.io/v1beta1/namespaces/openshift-machine-api/machinesets/openshift-blog-txxtf-worker-us-east-2a",

> 

Reviewer 3: remove lines 125-127.  No need to modify the self link, it will be updated when the definition json file is applied.

21c21

< "machine.openshift.io/cluster-api-machineset": "openshift-blog-txxtf-gpu-us-east-2a" --- > "machine.openshift.io/cluster-api-machineset": "openshift-blog-txxtf-worker-us-east-2a"

31c31

< "machine.openshift.io/cluster-api-machineset": "openshift-blog-txxtf-gpu-us-east-2a" --- > "machine.openshift.io/cluster-api-machineset": "openshift-blog-txxtf-worker-us-east-2a"

60c60

> < "instanceType": "p3.2xlarge", --- > "instanceType": "m4.large",

Reviewer 3: replace "p3.2xlarge" with "g4dn.xlarge"

104a105,111

> },

> "status": {

> "availableReplicas": 1,

> "fullyLabeledReplicas": 1,

> "observedGeneration": 1,

> "readyReplicas": 1,

> > "replicas": 1

Reviewer 3: Remove lines 141-155.  They are not needed.

----

. Create the new MachineSet from the definition:
+
[source,terminal]
----
$ oc create -f openshift-blog-txxtf-gpu-us-east-2a.json
----
+
.Example output
+
[source,terminal]
----
machineset.machine.openshift.io/openshift-blog-txxtf-gpu-us-east-2a created
----

. View the new MachineSet:
+
[source,terminal]
----
$ oc -n openshift-machine-api get machinesets | grep gpu
----
+
The MachineSet replica count is set to ``1`` so a new Machine object is created automatically.
+
.Example output
+
[source,terminal]
----
openshift-blog-txxtf-gpu-us-east-2a 1 1 1 1 4m21s
----

. View the new Machine object:
+
[source,terminal]
----
$ oc -n openshift-machine-api get machines | grep gpu
----
+
.Example output
+
[source,terminal]
----
openshift-blog-txxtf-gpu-us-east-2a-rd665 running p3.2xlarge us-east-2 us-east-2a 4m36s
----

. Eventually a node will come up based on the new definition. Find the node name.
+
[source,terminal]
----
$ oc -n openshift-machine-api get machines | grep gpu
----
+
.Example output
+
[source,terminal]
----

> openshift-blog-txxtf-gpu-us-east-2a-rd665 running p3.2xlarge us-east-2 us-east-2a 5m8s

Reviewer 3: replace "p3.2xlarge" with "g4dn.xlarge"

----

. View the metadata associated with the new node, including its labels. You can see the instance-type, operating system, zone, region, and hostname.
+
[source,terminal]
----
$ oc get node ip-10-0-132-138.us-east-2.compute.internal -o json | jq .metadata.labels
----
+
.Example output
+
[source,terminal]
----
{

"beta.kubernetes.io/arch": "amd64",

> "beta.kubernetes.io/instance-type": "p3.2xlarge",

Reviewer 3: replace "p3.2xlarge" with "g4dn.xlarge"
